### PR TITLE
Support for internal balance transfer

### PR DESCRIPTION
**Pull request by on247**
Implemented Tipping
Added new table transfers to schema
Changed sidebar to add transfer sections
Added views for Transfer history and requests
Modfied database.js 
 - Added support for querying and making transfers ( makeTransfer , getTransfers)

### CODE REVIEW

**Comment by RHavar** [comment]
Looking very good! Some changes:

Can you collapse this all into a nice single, clean commit, if all the lines are re-indented it makes it very difficult for me to review. Almost every line has changed :(

2. Don't add .gitignore to .gitignore file

In some of the error cases, you write:
```
if(err){
     callback("SENDER_DOES_NOT_EXIST");
    }

The problem with this, is execution will continue to flow through (resulting in a very serious bug)  (you need to add a `return`


I'll take a deeper look when the diff is more readable =)

**Comment by on247** [comment]
Fixed the missing returns
Now the code is readable

File: npm-debug.log [hunk 1]
```diff
@@ -0,0 +1,8 @@
+0 info it worked if it ends with ok
```
Comment (RHavar) [inline npm-debug.log]: This file shouldn't be commit, you can `git rm npm-debug.log`

File: server/database.js [hunk 2]
```diff
@@ -603,6 +622,52 @@ exports.getPublicStats = function(username, callback) {
...
+    // Update balances
+    getClient(function(client, callback) {
+      client.query("UPDATE users SET balance_satoshis = balance_satoshis - $1 WHERE id = $2",
+      [satoshis,senderId],function(err,response){
+
+        if (err)
+          return callback(err)
+
+        if (response.rowCount !== 1)
+        return callback(new Error('Unexpected sender balance update row count: \n' + response));
```
Comment (RHavar) [inline server/database.js]: Try not put a \n in here, because it makes it hard/impossible to grep

File: server/database.js [hunk 3]
```diff
@@ -603,6 +622,52 @@ exports.getPublicStats = function(username, callback) {
...
+
+        client.query("UPDATE users SET balance_satoshis = balance_satoshis + $1 WHERE id = $2",
+        [satoshis,destId],function(err,response){
+
+          if (err)
+            return callback(err)
+
+          if (response.rowCount !== 1)
+            return callback(new Error('Unexpected dest balance row count: \n' + response));
```
Comment (RHavar) [inline server/database.js]: try not put a \n in errors

File: server/database.js [hunk 4]
```diff
@@ -603,6 +622,52 @@ exports.getPublicStats = function(username, callback) {
...
+
+          client.query("INSERT INTO transfers (id,sender_user_id,dest_user_id,amount) values($1,$2,$3,$4) ",[uid,senderId,destId,satoshis],function(err,response){
+            if (err){
+              console.log(err);
```
Comment (RHavar) [inline server/database.js]: Don't console.log errors, there should be an internal errors. There should be a function in the common library for this, or use  `console.error('[INTERNAL ERROR] unable to insert into transfers: ', err);`

File: server/sql/schema.sql [hunk 5]
```diff
@@ -38,7 +39,32 @@ ALTER SEQUENCE users_id_seq OWNED BY users.id;
...
+ALTER TABLE ONLY transfers ADD CONSTRAINT transfers_pkey PRIMARY KEY (id);
+ALTER TABLE ONLY transfers ADD CONSTRAINT sender_user_id_fkey FOREIGN KEY (sender_user_id) REFERENCES   users(id) ON UPDATE CASCADE ON DELETE CASCADE;
+ALTER TABLE ONLY transfers ADD CONSTRAINT dest_user_id_fkey FOREIGN KEY (dest_user_id) REFERENCES   users(id) ON UPDATE CASCADE ON DELETE CASCADE;
+
+-- View that shows transfers with received and sent to usernames
 
+CREATE VIEW transfers_sent AS
+	(SELECT transfers.id , users.username as sender_username , transfers.created , amount from users
+		Inner join transfers on users.id = transfers.sender_user_id );
+CREATE VIEW transfers_received AS
```
Comment (RHavar) [inline server/sql/schema.sql]: I don't think you should be using views, but the natural join you use doesn't look right

File: server/database.js [hunk 6]
```diff
@@ -655,6 +720,28 @@ exports.getWithdrawals = function(userId, callback) {
...
 };
 
+exports.getTransfers =function (userId,callback){
+  assert(userId && callback);
+  getUserFromId(userId,function(err,sender){
+    if(err){
+      return callback("SENDER_DOES_NOT_EXIST");
+    }
+    var username=sender.username;
+    query("select * from transfers_received_sent where sender_username = $1 or dest_username = $1 ORDER BY created DESC" ,[username] ,function(err,result){
```
Comment (RHavar) [inline server/database.js]: The index is on LOWER(username)  so you'll need to write:

LOWER(sender_username) = LOWER($1)

etc   (Also it makes it case-insensitive)

File: .gitignore [hunk 7]
```diff
@@ -1,7 +1,6 @@
+
```
Comment (RHavar) [inline .gitignore]: try not modify files you aren't modifying :D

File: server/sql/schema.sql [hunk 8]
```diff
@@ -38,7 +39,32 @@ ALTER SEQUENCE users_id_seq OWNED BY users.id;
...
 ALTER TABLE ONLY users ALTER COLUMN id SET DEFAULT nextval('users_id_seq'::regclass);
 
+-- Transfers (Tips)
+CREATE TABLE transfers (
+  id uuid NOT NULL,
+  created timestamp with time zone DEFAULT now() NOT NULL,
+  sender_user_id bigint NOT NULL,
+  dest_user_id bigint NOT NULL,
+  amount bigint NOT NULL,
+  CONSTRAINT user_transfer_valid_amount CHECK(amount!=0)
```
Comment (RHavar) [inline server/sql/schema.sql]: should be CHECK(amount>0)

File: server/user.js [hunk 9]
```diff
@@ -624,6 +625,104 @@ exports.deposit = function(req, res, next) {
...
+  **/
+
+ exports.handleTransferRequest = function (req,res,next){
+     var uid = req.body.transfer_id
+     var user = req.user;
+     var amount = req.body.amount;
+     var destUser = req.body.destUser;
+     var password = lib.removeNullsAndTrim(req.body.password);
+     var otp = lib.removeNullsAndTrim(req.body.otp);
+     var r =  /^[1-9][0-9]*$/;
```
Comment (RHavar) [inline server/user.js]: You should copy the code from handleWithdrawalRequest, it should allow sending fractions of a  bit (so you can tip your whole balance)

File: server/user.js [hunk 10]
```diff
@@ -624,6 +625,104 @@ exports.deposit = function(req, res, next) {
...
+     if (!r.test(amount))
+         return res.render('transfer-request', { user: user, id: uuid.v4(),  warning: 'Not a valid amount' });
+    amount = Math.round(parseFloat(amount) * 100)
```
Comment (RHavar) [inline server/user.js]: You should put a min tip of 100 bits

File: server/user.js [hunk 11]
```diff
@@ -12,6 +12,7 @@ var qr = require('qr-image');
 var uuid = require('uuid');
 var _ = require('lodash');
 var config = require('../config/config');
+var util=require('util');
```
Comment (RHavar) [inline server/user.js]: ?

File: server/user.js [hunk 12]
```diff
@@ -624,6 +625,104 @@ exports.deposit = function(req, res, next) {
...
+            if (err === 'INVALID_OTP')
+                return res.render('transfer-request', { user: user, id: uuid.v4(), warning: 'invalid one-time token' });
+            //Should be an user
+            return next(new Error('Unable to validate user handling transfer: \n' + err));
+        }
+        // Check destination user
+
+        database.getUserFromUsername(destUser,function (err,userData){
+            if(err){
+              if(err="NO_USER"){
```
Comment (RHavar) [inline server/user.js]: This shouldn't be using =, it should be using equality

File: server/user.js [hunk 13]
```diff
@@ -624,6 +625,104 @@ exports.deposit = function(req, res, next) {
...
+                  return res.render('transfer-request', { user: user, id: uuid.v4(), warning: 'The user you are sending to does not exist' });
+              }
```
Comment (RHavar) [inline server/user.js]: You don't handle the case of an error not being "NO_USER"

File: server/user.js [hunk 14]
```diff
@@ -624,6 +625,104 @@ exports.deposit = function(req, res, next) {
...
+            }
+            var destUserId=userData.id;
+            var senderUserId=req.user.id;
+            if(senderUserId==destUserId){
+                return res.render('transfer-request', { user: user, id: uuid.v4(), warning: 'You cant send to yourself ...' });
+            }
+
+            database.makeTransfer(uid,senderUserId,destUserId,amount,function(err){
+              if(err){
+                console.log(err);
```
Comment (RHavar) [inline server/user.js]: Don't console.log errors

File: server/user.js [hunk 15]
```diff
@@ -624,6 +625,104 @@ exports.deposit = function(req, res, next) {
...
+                if(err == "NOT_ENOUGH_MONEY"){
+                    return res.render('transfer-request', { user: user, id: uuid.v4(), warning: 'You dont have enough bits to transfer' });
+                }
+                //Should never happen
+                else {
```
Comment (RHavar) [inline server/user.js]: Never silently ignore stuff. Log an internal error with the details

**Comment by RHavar** [comment]
Closing, as I implemented in 3f224a728fd2eed3c49dcd479fc7c2d20cc6b38c

**Pull request closed without merging**